ui-v2/app/initializers/nspace.js

```diff
@@ -0,0 +1,61 @@
+import Route from '@ember/routing/route';
+import config from 'consul-ui/config/environment';
+
+import { routes } from 'consul-ui/router';
+import flat from 'flat';
+
+let initialize = function() {};
+Route.reopen(
+  ['modelFor', 'transitionTo', 'replaceWith', 'paramsFor'].reduce(function(prev, item) {
+    prev[item] = function(routeName, ...rest) {
+      const isNspaced = this.routeName.startsWith('nspace.');
+      if (routeName === 'nspace') {
+        if (isNspaced || this.routeName === 'nspace') {
+          return this._super(...arguments);
+        } else {
+          return {
+            nspace: '~',
+          };
+        }
+      }
+      if (isNspaced && routeName.startsWith('dc')) {
+        return this._super(...[`nspace.${routeName}`, ...rest]);
+      }
+      return this._super(...arguments);
+    };
+    return prev;
+  }, {})
+);
+if (config.CONSUL_NSPACES_ENABLED) {
+  const dotRe = /\./g;
+  initialize = function(container) {
+    const all = Object.keys(flat(routes))
+      .filter(function(item) {
+        return item.startsWith('dc');
+      })
+      .map(function(item) {
+        return item.replace('._options.path', '').replace(dotRe, '/');
+      });
+    all.forEach(function(item) {
+      let route = container.resolveRegistration(`route:${item}`);
+      if (!route) {
+        item = `${item}/index`;
+        route = container.resolveRegistration(`route:${item}`);
+      }
+      route.reopen({
+        templateName: item
+          .replace('/root-create', '/create')
+          .replace('/create', '/edit')
+          .replace('/folder', '/index'),
+      });
+      container.register(`route:nspace/${item}`, route);
+      const controller = container.resolveRegistration(`controller:${item}`);
+      if (controller) {
+        container.register(`controller:nspace/${item}`, controller);
+      }
+    });
+  };
```

Comment:
This ends up doubling a bunch of registrations - have you seen any issues with that? I _think_ there's no way to get to the non-namespaces registrations via normal app interfaces since you're re-writing route names for links, but didn't know if it'd cause other weirdness.

Reply:
Nope not seen any issues. If anything I'm guessing we could get issues with sticky queryParams here (it may stick to both), but we aren't use those anywhere as yet.

> I think there's no way to get to the non-namespaces registrations via normal app interfaces since you're re-writing route names for links

From what I remember we only rewrite the route names if you are in a namespace, if you aren't you just get normal routes.